projects/crash_ping_delays_pingSender.kp/knowledge.md

```diff
@@ -0,0 +1,323 @@
+                                              "meta/creationTimestamp",
+                                              "meta/Date",
+                                              "meta/Timestamp",
+                                              "payload/crashDate"])
+
+post_subset = get_pings_properties(post_pings, ["application/channel",
+                                               "id",
+                                               "payload/processType",
+                                               "creationDate",
+                                               "meta/creationTimestamp",
+                                               "meta/Date",
+                                               "meta/Timestamp",
+                                               "payload/crashDate"])
+```
+The rest of the analysis is cleaner if we combine the two cohorts here.
+
+
+```python
+def add_pre(p):
+    p['pre'] = 'pre'
+    return p
+
+def add_post(p):
+    p['pre'] = 'post'
+    return p
+
+combined = pre_subset.map(add_pre).union(post_subset.map(add_post))
+```
+Quick normalization: ditch any ping that doesn't have a subsessionLength, creationTimestamp, or Timestamp:
+
+
+```python
+prev_count = combined.count()
+combined = combined.filter(lambda p:\
+                       p["payload/crashDate"] is not None\
+                       and p["meta/Timestamp"] is not None\
+                       and p["meta/creationTimestamp"] is not None)
+filtered_count = combined.count()
+print "Filtered {} of {} pings ({:.2f}%)".format(prev_count - filtered_count, prev_count, 100.0 * (prev_count - filtered_count) / prev_count)
+```
+    Filtered 0 of 186321 pings (0.00%)
+
+
+`pingSender` only submits "crash" pings for main-process crashes, so let's limit ourselves to those.
+
+
+```python
+prev_count = combined.count()
+combined = combined.filter(lambda p: p["payload/processType"] == "main")
+filtered_count = combined.count()
+print "Filtered {} of {} pings ({:.2f}%)".format(prev_count - filtered_count, prev_count, 100.0 * (prev_count - filtered_count) / prev_count)
+```
+    Filtered 138409 of 186321 pings (74.29%)
+
+
+##### Deduplication
+We sometimes receive crash pings more than once (identical document ids). This is usually below 1%, but there was a [known issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1345153) that can complicate measurement.
+
+So we'll dedupe here.
+
+
+```python
+combined_deduped = combined\
+    .map(lambda p: (p["id"], p))\
+    .reduceByKey(lambda a, b: a if a["meta/Timestamp"] < a["meta/Timestamp"] else b)\
+    .map(lambda pair: pair[1])
+```
+
+```python
+combined_count = combined.count()
+combined_deduped_count = combined_deduped.count()
+print "Filtered {} of {} crash pings ({:.2f}%)".format(combined_count - combined_deduped_count, combined_count, 100.0 * (combined_count - combined_deduped_count) / combined_count)
+```
+    Filtered 9684 of 47912 crash pings (20.21%)
```

Comment:
Can you add a note summarizing the cumulative impact of the deduplication and main process filtering? E.g. pingSender affects ~20% of all crash pings.